pkg/agent/apiserver/handlers/ovsflows/handler_test.go

```diff
@@ -171,43 +172,126 @@ func TestServiceFlows(t *testing.T) {
+			query:          "?networkpolicy=np1&namespace=default&type=K8sNP",
 			expectedStatus: http.StatusOK,
 		},
 		{
 			test:           "Non-existing NetworkPolicy",
 			name:           "np2",
 			namespace:      "ns2",
-			query:          "?networkpolicy=np2&&namespace=ns2",
+			policyType:     cpv1beta.K8sNetworkPolicy,
+			query:          "?networkpolicy=np2&namespace=ns2&type=K8sNP",
 			expectedStatus: http.StatusNotFound,
 		},
+		{
+			test:           "Ambiguous query",
+			name:           "np3",
+			namespace:      "ns3",
+			query:          "?networkpolicy=np3&namespace=ns3",
+			expectedStatus: http.StatusBadRequest,
+			expectedErr:    errAmbiguousQuery,
+		},
+		{
+			test:           "Existing ACNP",
+			name:           "acnp1",
+			policyType:     cpv1beta.AntreaClusterNetworkPolicy,
+			query:          "?networkpolicy=acnp1&type=ACNP",
+			expectedStatus: http.StatusOK,
+		},
+		{
+			test:           "ACNP bad request",
+			name:           "acnp2",
+			policyType:     cpv1beta.AntreaClusterNetworkPolicy,
+			query:          "?networkpolicy=acnp2&type=ACNP&namespace=default",
+			expectedStatus: http.StatusBadRequest,
+		},
+		{
+			test:           "Existing ANNP",
+			name:           "annp1",
+			namespace:      "default",
+			policyType:     cpv1beta.AntreaNetworkPolicy,
+			query:          "?networkpolicy=annp1&namespace=default&type=ANNP",
+			expectedStatus: http.StatusOK,
+		},
+		{
+			test:           "ANNP bad request",
+			name:           "annp2",
+			policyType:     cpv1beta.AntreaNetworkPolicy,
+			query:          "?networkpolicy=annp2&type=ANNP",
+			expectedStatus: http.StatusBadRequest,
+		},
+		{
+			test:           "Existing ANP",
+			name:           "anp1",
+			policyType:     cpv1beta.AdminNetworkPolicy,
+			query:          "?networkpolicy=anp1&type=ANP",
+			expectedStatus: http.StatusOK,
+		},
 	}
 	for i := range testcases {
 		tc := testcases[i]
 		npq := queriertest.NewMockAgentNetworkPolicyInfoQuerier(ctrl)
 		q := aqtest.NewMockAgentQuerier(ctrl)
-		q.EXPECT().GetNetworkPolicyInfoQuerier().Return(npq).Times(1)
-
-		if tc.expectedStatus != http.StatusNotFound {
-			ofc := oftest.NewMockClient(ctrl)
-			ovsctl := ovsctltest.NewMockOVSCtlClient(ctrl)
-			npq.EXPECT().GetNetworkPolicies(&querier.NetworkPolicyQueryFilter{SourceName: tc.name, Namespace: tc.namespace}).Return([]cpv1beta.NetworkPolicy{*testNetworkPolicy}).Times(1)
-			ofc.EXPECT().GetNetworkPolicyFlowKeys(tc.name, tc.namespace).Return(testFlowKeys).Times(1)
-			q.EXPECT().GetOpenflowClient().Return(ofc).Times(1)
-			q.EXPECT().GetOVSCtlClient().Return(ovsctl).Times(len(testFlowKeys))
-			for i := range testFlowKeys {
-				ovsctl.EXPECT().DumpMatchedFlow(testFlowKeys[i]).Return(testDumpFlows[i], nil).Times(1)
+		npFilter := &querier.NetworkPolicyQueryFilter{
+			SourceName: tc.name,
+			Namespace:  tc.namespace,
+			SourceType: tc.policyType,
+		}
+		if tc.expectedStatus != http.StatusBadRequest || tc.expectedErr == errAmbiguousQuery {
```

Comment:
I don't feel super strongly about it, but it my view we should just define expectations as part of the test case definition, like we often do in our unit tests. This would avoid all the branching in the test definition, which itself depends on the expected test outcome. For example:
```go
		{
			test:           "Ambiguous query",
			query:          "?networkpolicy=np1&namespace=ns1",
			expectedStatus: http.StatusBadRequest,
			expectedErr:    errAmbiguousQuery,
			expectedCalls: func(ctrl *gomock.Controller) {
			        npq := queriertest.NewMockAgentNetworkPolicyInfoQuerier(ctrl)
			        q := aqtest.NewMockAgentQuerier(ctrl)
			        q.EXPECT().GetNetworkPolicyInfoQuerier().Return(npq)
			        npq.EXPECT().GetNetworkPolicies(&querier.NetworkPolicyQueryFilter{SourceName: "np1", Namespace: "ns1"}).Return([]cpv1beta.NetworkPolicy{cpv1beta.NetworkPolicy{}, cpv1beta.NetworkPolicy{}})
			},
		},
		{
			test:           "Existing ACNP",
			query:          "?networkpolicy=acnp1&type=ACNP",
			expectedStatus: http.StatusOK,
			expectedCalls: expectedCallsSuccess("acnp1", "", cpv1beta.AntreaClusterNetworkPolicy),
		},
```

with:
```go
func expectedCallsSuccess(name, namespace string, policyType cpv1beta.NetworkPolicyType) func(ctrl *gomock.Controller) {
	npFilter := &querier.NetworkPolicyQueryFilter{
	        SourceName: name,
		Namespace:  namespace,
		SourceType: policyType,
	}
        return func(ctrl *gomock.Controller) {
                npq := queriertest.NewMockAgentNetworkPolicyInfoQuerier(ctrl)
                q := aqtest.NewMockAgentQuerier(ctrl)
                ofc := oftest.NewMockClient(ctrl)
		ovsctl := ovsctltest.NewMockOVSCtlClient(ctrl)
                q.EXPECT().GetNetworkPolicyInfoQuerier().Return(npq)
                npq.EXPECT().GetNetworkPolicies(npFilter).Return([]cpv1beta.NetworkPolicy{
                        cpv1beta.NetworkPolicy{
		                SourceRef: &cpv1beta.NetworkPolicyReference{
			                Type:      policyType,
			                Namespace: namespace,
			                Name: name,
		                },
	                },
                })
		ofc.EXPECT().GetNetworkPolicyFlowKeys(name, namespace, policyType).Return(testFlowKeys)
		q.EXPECT().GetOpenflowClient().Return(ofc).Times(1)
		q.EXPECT().GetOVSCtlClient().Return(ovsctl).Times(len(testFlowKeys))
		for i := range testFlowKeys {
		        ovsctl.EXPECT().DumpMatchedFlow(testFlowKeys[i]).Return(testDumpFlows[i], nil)
		}
}
```

Reply:
I agree that we should not use the expected status as branch conditions. However, adding expected calls in the testcases here are much trickier, and by the end of the day the `expectedCallsSuccess` need to return essentially all mocked instances to be passed to the `runHttpTest` function, which is shared by other tests. I don't feel like it's worth the complexity.
The easiest solution I figured was to simply group tests with different expected return status into separate test functions. It is much cleaner to read and easier to debug.